collector/extension/memorylimiterextension/memorylimiter.go

```diff
@@ -0,0 +1,165 @@
+	"context"
+	"fmt"
+	"time"
+
+	arrowpb "github.com/open-telemetry/otel-arrow/api/experimental/arrow/v1"
+	"golang.org/x/sync/semaphore"
+	"go.uber.org/zap"
+	"github.com/golang/protobuf/proto"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+
+	"go.opentelemetry.io/collector/component"
+)
+
+type wrappedHandlerFn func(ctx context.Context, req any, handler grpc.UnaryHandler) (any, error)
+type memoryLimiterExtension struct {
+	limitBytes int64
+	sem        *semaphore.Weighted
+	logger     *zap.Logger
+	timeout    time.Duration
+	UnaryHandlerFn wrappedHandlerFn 
+}
+
+type requestWrappedServerStream struct {
+	wrappedCtx context.Context
+	stream grpc.ServerStream
+	pendingReqs chan any 
+	errCh chan error
+	timeout time.Duration
+}
+
+func (rss *requestWrappedServerStream) SetHeader(md metadata.MD) error {
+	return rss.stream.SetHeader(md)
+}
+
+func (rss *requestWrappedServerStream) SendHeader(md metadata.MD) error {
+	return rss.stream.SendHeader(md)
+}
+
+func (rss *requestWrappedServerStream) SetTrailer(md metadata.MD) {
+	rss.stream.SetTrailer(md)
+}
+
+func (rss *requestWrappedServerStream) Context() context.Context {
+	return rss.stream.Context()
+}
+
+func (rss *requestWrappedServerStream) SendMsg(m any) error {
+	return rss.stream.SendMsg(m)
+}
+
+func (rss *requestWrappedServerStream) RecvMsg(m any) error {
+	// return rss.stream.RecvMsg(m)
+	fmt.Println("ENTERED RECVMSG")
+
+	select {
+	case a := <-rss.pendingReqs:
+		src := a.(*arrowpb.BatchArrowRecords)
+		dest := m.(*arrowpb.BatchArrowRecords)
+		*dest = *src 
+	case err := <-rss.errCh:
+		return err
+	}
+	return nil
+}
+
+// This interface is meant to access the size of a
+// ExportTraceServiceRequest, ExportMetricsServiceRequest, ExportLogsServicesRequest
+type telemetryServiceRequest = interface { 
+	Size() int
+}
+
+func (ml *memoryLimiterExtension) UnaryInterceptorGenerator() grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req any, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp any, err error) {
+		a := req.(telemetryServiceRequest)
+		requestSize := int64(a.Size())
+		
+		semCtx, cancel := context.WithTimeout(context.Background(), ml.timeout)
+		defer cancel()
+
+		err = ml.sem.Acquire(semCtx, requestSize)
+		if err != nil {
+			return nil, fmt.Errorf("not enough memory available to process request, %w", err)
+		}
+
+		resp, err = handler(ctx, req)
+		ml.sem.Release(requestSize)
+
+		return resp, err
+	}
+}
+
+type streamTelemetryServiceRequest = interface { 
+	Size() int
+	arrowpb.BatchArrowRecords
+}
+func (ml *memoryLimiterExtension) StreamInterceptorGenerator() grpc.StreamServerInterceptor {
+	return func(srv any, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		rss := &requestWrappedServerStream{
+			stream: ss,
+			pendingReqs: make(chan any, 5),
+			errCh: make(chan error, 5),
+			timeout: ml.timeout,
+		}
+		a := new(arrowpb.BatchArrowRecords)
+		var err error
+		err = ss.RecvMsg(a)
```

Comment:
Calls RecvMsg() on the underlying stream to see the request so that we can find its size. Note that the receiver will also call Recv() so we need to wrap the underlying stream with a custom `wrappedStream.RecvMsg()` that the receiver will call.